src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -58,6 +59,93 @@ def check_type_refinement(node: foast.Expr, new: foast.DataType) -> None:
             )
 
 
+@dataclass
+class TypeInfo:
+
+    type: foast.SymbolType  # noqa: A003
+
+    @property
+    def is_complete(self) -> bool:
+        return is_complete_symbol_type(self.type)
+
+    @property
+    def is_any_type(self) -> bool:
+        if self.is_complete:
+            return False
+        return (self.type is None) or (self.type.constraint is None)
+
+    @property
+    def constraint(self) -> Optional[foast.SymbolType]:
+        if self.is_complete:
+            return self
+        elif not self.is_any:
+            return self.constraint
+
+    @property
+    def is_field_type(self) -> bool:
+        return isinstance(self.type, foast.FieldType) or self.constraint is foast.FieldType
+
+    @property
+    def is_scalar(self) -> bool:
+        if self.is_complete:
+            return isinstance(self.type, foast.ScalarType)
+        else:
+            return isinstance(self.constraint, foast.ScalarType)
+        return False
+
+    @property
+    def is_arithmetic_compatible(self) -> bool:
+        match self.type:
+            case foast.FieldType(dtype=foast.ScalarType(kind=dtype_kind)) | foast.ScalarType(
+                kind=dtype_kind
+            ):
+                if dtype_kind is not foast.ScalarKind.BOOL:
+                    return True
+        return False
+
+    @property
+    def is_logics_compatible(self) -> bool:
+        match self.type:
+            case foast.FieldType(dtype=foast.ScalarType(kind=dtype_kind)) | foast.ScalarType(
+                kind=dtype_kind
+            ):
+                if dtype_kind is foast.ScalarKind.BOOL:
+                    return True
+        return False
+
+    def can_be_refined_to(self, other: "TypeInfo") -> bool:
+        if self.is_any_type:
+            return True
+        if self.is_complete:
+            return False
+        if self.constraint:
+            if other.is_complete:
+                return isinstance(other.type, self.constraint)
+            elif other.constraint:
+                return self.constraint is other.constraint
+        return False
+
+
+def dims_promotable(left: TypeInfo, right: TypeInfo) -> bool:
```

Comment:
It's not critical but I think it's more appropriate use  the "broadcast" instead of "promotion" when talking about dimensions and not types.

Reply:
agreed.